File: packages/jetbrains-plugin/src/test/java/com/chainlesschain/ide/ApprovalSettlementRegistryTest.java
package com.chainlesschain.ide;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;
import org.junit.jupiter.api.Test;

class ApprovalSettlementRegistryTest {

    @Test
    @SuppressWarnings("unchecked")
    void replaysJetbrainsSurfaceFromSharedHumanTaskFixture() throws Exception {
        Map<String, Object> fixture = (Map<String, Object>) MiniJson.parse(
                Files.readString(fixturePath(), StandardCharsets.UTF_8));
        List<Map<String, Object>> scenarios =
                (List<Map<String, Object>>) (List<?>) fixture.get("scenarios");
        int replayed = 0;

        for (Map<String, Object> scenario : scenarios) {
            List<String> surfaces = (List<String>) (List<?>) scenario.get("surfaces");
            if (!surfaces.contains("jetbrains")) continue;
            replayed += 1;

            String name = String.valueOf(scenario.get("name"));
            String requestId = "approval-" + name;
            ApprovalSettlementRegistry registry = new ApprovalSettlementRegistry();
            assertTrue(registry.open(requestId), name);
            int sentDecisions = 0;
            int interrupts = 0;
            int rejectedResponses = 0;
            String unresolvedDecision = null;

            List<Map<String, Object>> steps =
                    (List<Map<String, Object>>) (List<?>) scenario.get("steps");
            for (Map<String, Object> step : steps) {
                String action = String.valueOf(step.get("action"));
                Map<String, Object> expectations =
                        (Map<String, Object>) step.get("expect");
                String expectedStep = String.valueOf(expectations.get("jetbrains"));

                if ("restart".equals(action)) {
                    if (unresolvedDecision != null) {
                        registry.resolve(requestId);
                        unresolvedDecision = null;
                    }
                    registry.invalidateAll();
                    assertEquals("settled", expectedStep, name);
                    continue;
                }

                if ("cancel".equals(action)) {
                    List<String> reserved = registry.beginInterrupt();
                    interrupts += 1;
                    for (String id : reserved) {
                        assertTrue(registry.complete(
                                id, ApprovalSettlementRegistry.Status.INTERRUPTING, true), name);
                    }
                    assertEquals("settled", expectedStep, name);
                    continue;
                }

                assertTrue("approve".equals(action) || "decline".equals(action),
                        name + ": unsupported JetBrains action " + action);
                boolean accepted = registry.beginDecision(requestId);
                if (accepted) {
                    assertTrue(registry.complete(
                            requestId, ApprovalSettlementRegistry.Status.RESPONDING, true), name);
                    sentDecisions += 1;
                    unresolvedDecision = action;
                } else {
                    rejectedResponses += 1;
                }
                assertEquals("settled".equals(expectedStep), accepted, name);
            }

            if (unresolvedDecision != null) registry.resolve(requestId);
            Map<String, Object> expectedBySurface =
                    (Map<String, Object>) scenario.get("expected");
            Map<String, Object> expected =
                    (Map<String, Object>) expectedBySurface.get("jetbrains");
            assertEquals(asInt(expected.get("pending_approvals")), registry.size(), name);
            assertEquals(asInt(expected.get("sent_decisions")), sentDecisions, name);
            assertEquals(asInt(expected.get("interrupts")), interrupts, name);
            assertEquals(asInt(expected.get("rejected_responses")), rejectedResponses, name);
        }

        assertEquals(4, replayed, "JetBrains fixture coverage changed unexpectedly");
    }

    @Test
    void rejectedTransportRollsReservationBackForRetry() {
        ApprovalSettlementRegistry registry = new ApprovalSettlementRegistry();
        assertTrue(registry.open("approval-retry"));
        assertTrue(registry.beginDecision("approval-retry"));
        assertTrue(registry.complete(
                "approval-retry", ApprovalSettlementRegistry.Status.RESPONDING, false));
        assertEquals(ApprovalSettlementRegistry.Status.PENDING,
                registry.status("approval-retry"));
        assertTrue(registry.beginDecision("approval-retry"));
    }

    @Test
    void concurrentDecisionReservationHasExactlyOneWinner() {
        ApprovalSettlementRegistry registry = new ApprovalSettlementRegistry();
        assertTrue(registry.open("approval-race"));
        AtomicInteger winners = new AtomicInteger();

        IntStream.range(0, 64).parallel().forEach(ignored -> {
            if (registry.beginDecision("approval-race")) winners.incrementAndGet();
        });

        assertEquals(1, winners.get());
        assertFalse(registry.beginDecision("approval-race"));
    }

    private static int asInt(Object value) {
        return ((Number) value).intValue();
    }

    private static Path fixturePath() {
        return Path.of("..", "agent-protocol", "test", "fixtures",
                "human-task-settlement-conformance.json");
    }
}

File: android-app/core-agent-protocol/src/test/java/com/chainlesschain/android/core/agentprotocol/ApprovalSettlementRegistryTest.kt
package com.chainlesschain.android.core.agentprotocol

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ApprovalSettlementRegistryTest {
    @Test
    fun `replays Android surface from shared HumanTask fixture`() {
        val fixture = javaClass.classLoader!!
            .getResourceAsStream("human-task-settlement-conformance.json")!!
            .bufferedReader()
            .use { Json.parseToJsonElement(it.readText()).jsonObject }
        var replayed = 0
        for (scenarioValue in fixture.getValue("scenarios").jsonArray) {
            val scenario = scenarioValue.jsonObject
            val surfaces = scenario.getValue("surfaces").jsonArray
                .map { it.jsonPrimitive.content }
            if ("android" !in surfaces) continue
            replayed += 1
            val name = scenario.getValue("name").jsonPrimitive.content
            val requestId = "approval-$name"
            val registry = ApprovalSettlementRegistry()
            assertTrue(name, registry.open(requestId))
            var sentDecisions = 0
            var interrupts = 0
            var rejectedResponses = 0
            var unresolvedDecision = false

            for (stepValue in scenario.getValue("steps").jsonArray) {
                val step = stepValue.jsonObject
                val action = step.getValue("action").jsonPrimitive.content
                val expected = step.getValue("expect").jsonObject
                    .getValue("android").jsonPrimitive.content
                when (action) {
                    "restart" -> {
                        if (unresolvedDecision) registry.resolve(requestId)
                        unresolvedDecision = false
                        registry.invalidateAll()
                        assertEquals(name, "settled", expected)
                    }
                    "cancel" -> {
                        val reserved = registry.beginInterrupt()
                        interrupts += 1
                        reserved.forEach {
                            assertTrue(
                                name,
                                registry.complete(
                                    it,
                                    ApprovalSettlementRegistry.Status.INTERRUPTING,
                                    true,
                                ),
                            )
                        }
                        assertEquals(name, "settled", expected)
                    }
                    "approve", "decline" -> {
                        val accepted = registry.beginDecision(requestId)
                        if (accepted) {
                            assertTrue(
                                name,
                                registry.complete(
                                    requestId,
                                    ApprovalSettlementRegistry.Status.RESPONDING,
                                    true,
                                ),
                            )
                            sentDecisions += 1
                            unresolvedDecision = true
                        } else {
                            rejectedResponses += 1
                        }
                        assertEquals(name, expected == "settled", accepted)
                    }
                    else -> error("unsupported Android action $action")
                }
            }
            if (unresolvedDecision) registry.resolve(requestId)
            val expected = scenario.getValue("expected").jsonObject
                .getValue("android").jsonObject
            assertEquals(name, expected.int("pending_approvals"), registry.size())
            assertEquals(name, expected.int("sent_decisions"), sentDecisions)
            assertEquals(name, expected.int("interrupts"), interrupts)
            assertEquals(name, expected.int("rejected_responses"), rejectedResponses)
        }
        assertEquals(4, replayed)
    }

    @Test
    fun `failed send rolls back and concurrent decision has one winner`() {
        val registry = ApprovalSettlementRegistry()
        assertTrue(registry.open("approval-race"))
        val winners = java.util.concurrent.atomic.AtomicInteger()
        val threads = List(64) {
            Thread {
                if (registry.beginDecision("approval-race")) winners.incrementAndGet()
            }
        }
        threads.forEach(Thread::start)
        threads.forEach(Thread::join)
        assertEquals(1, winners.get())
        assertTrue(
            registry.complete(
                "approval-race",
                ApprovalSettlementRegistry.Status.RESPONDING,
                false,
            ),
        )
        assertEquals(
            ApprovalSettlementRegistry.Status.PENDING,
            registry.status("approval-race"),
        )
        assertTrue(registry.beginDecision("approval-race"))
        assertFalse(registry.open("approval-race"))
    }

    private fun kotlinx.serialization.json.JsonObject.int(name: String): Int =
        getValue(name).jsonPrimitive.content.toInt()
}
